kmk/modules/text_replacement.py

```diff
@@ -0,0 +1,202 @@
+try:
+    from typing import Optional
+except ImportError:
+    # we're not in a dev environment, so we don't need to worry about typing
+    pass
+from micropython import const
+
+from kmk.keys import KC, Key
+from kmk.modules import Module
+
+
+class State:
+    LISTENING = const(0)
+    DELETING = const(1)
+    SENDING = const(2)
+
+
+class Character:
+    '''Helper class for making a left-shifted key identical to a right-shifted key'''
+
+    is_shifted: bool = False
+
+    def __init__(self, key_code: Key, is_shifted: bool) -> None:
+        self.is_shifted = is_shifted
+        self.key_code = KC.LSHIFT(key_code) if is_shifted else key_code
+
+    def __eq__(self, other: any) -> bool:  # type: ignore
+        try:
+            return (
+                self.key_code.code == other.key_code.code
+                and self.is_shifted == other.is_shifted
+            )
+        except AttributeError:
+            return False
+
+
+class Phrase:
+    '''Manages a collection of characters and keeps an index of them so that potential matches can be tracked'''
+
+    def __init__(self, string: str) -> None:
+        self._characters: list[Character] = []
+        self._index: int = 0
+        for char in string:
+            try:
+                key_code = getattr(KC, char.upper())

```

Comment:
```suggestion
                key_code = KC[char]
```
`key.upper()` is also applied in `KC.__get_item__`